src/crypto/secretstream/secretstream_macros.rs

```diff
@@ -0,0 +1,354 @@
+    /// consumes the `Encryptor`, so it can no longer be used after a call to
+    /// `finalize`.
+    pub fn aencrypt_finalize(mut self, m: &[u8], ad: Option<&[u8]>) -> Result<Vec<u8>, ()> {
+        self.aencrypt(m, ad, TAG_FINAL)
+    }
+
+    /// This method explicitly re-keys the `Encryptor` and updates its state, but
+    /// doesn't add any information about the key change to the stream. If this
+    /// function is used to create an encrypted stream, the decryption process
+    /// must call that function at the exact same stream location.
+    /// See also the method `aencrypt_rekey`.
+    pub fn rekey(&mut self) {
+        unsafe {
+            $rekey_name(&mut self.0);
+        }
+    }
+}
+
+/// `Decryptor` contains the state for multi-part (streaming) computations. This
+/// allows the caller to process encryption of a sequence of multiple messages.
+/// After the last message of a valid stream containing the tag `Tag::Finalized`
+/// is decrypted, the `Decryptor` may no longer verify and decrypt messages or
+/// re-key itself.
+pub struct Decryptor {
+    state: $state_name,
+    finalized: bool,
+}
+
+impl Decryptor {
+    /// Initializes a `Decryptor` given a secret `Key` and a `Header`. The key
+    /// will not be required any more for subsequent operations. `Err(())` is
+    /// returned if the header is invalid.
+    pub fn init(header: &Header, key: &Key) -> Result<Self, ()> {
+        let mut state: $state_name = unsafe { mem::uninitialized() };
+
+        let rc = unsafe {
+            $init_pull_name(&mut state, header.0.as_ptr(), key.0.as_ptr())
+        };
+        if rc == -1 {
+            // NOTE: this return code explicitly means the header is invalid,
+            // but when implementing error types we should still consider the
+            // possibility of some other non-zero code below with a generic call
+            // to external function failed error.
+            return Err(());
+        } else if rc != 0 {
+            return Err(());
+        }
+
+        Ok(Self{state, finalized: false})
+    }
+
+    /// Verifies that `c` is a valid ciphertext with a correct authentication tag
+    /// given the internal state of the `Decryptor` (ciphertext streams cannot be
+    /// decrypted out of order for this reason). Also may validate the optional
+    /// unencrypted additional data `ad` using the authentication tag attached to
+    /// `c`. Finally decrypts the ciphertext and tag, and checks the tag
+    /// validity.
+    /// If any authentication fails, the stream has already been finalized, or if
+    /// the tag byte for some reason does not correspond to a valid `Tag`,
+    /// returns `Err(())`. Otherwise returns the plaintext and the tag.
+    /// Applications will typically use a `while decryptor.is_not_finalized()`
+    /// loop to authenticate and decrypt a stream of messages.
+    pub fn vdecrypt(&mut self, c: &[u8], ad: Option<&[u8]>) -> Result<(Vec<u8>, Tag),()> {
```

Comment:
This could be called ‘open()’

Reply:
Maybe we could consume self and return it here together with the tag, so that we don’t return self if the state is finalized. Then it won’t be possible for a user to try to keep using the state even though it has been finalized. That change would lead to the Tag enum having to contain the decrypter.